pkg/scheduler/cache/cache.go

```diff
@@ -176,6 +184,236 @@ type imageState struct {
+	Name           string
+	Tier           int
+	Parent         *HyperNodeTreeNode
+	Members        sets.Set[*HyperNodeTreeNode]
+	MemberMatchers sets.Set[*MemberMatcher]
+	Nodes          sets.Set[*schedulingapi.NodeInfo]
+	NodeMatchers   sets.Set[*MemberMatcher]
+	HyperNodeRef   *networktopov1alpha1.HyperNode
+}
+
+func NewHyperNodeTreeNode(name string) *HyperNodeTreeNode {
+	return &HyperNodeTreeNode{
+		Name:           name,
+		Tier:           -1, // Initialize to an invalid value
+		Members:        sets.New[*HyperNodeTreeNode](),
+		MemberMatchers: sets.New[*MemberMatcher](),
+		Nodes:          sets.New[*schedulingapi.NodeInfo](),
+		NodeMatchers:   sets.New[*MemberMatcher](),
+	}
+}
+
+type HyperNodeSet sets.Set[*HyperNodeTreeNode]
+
+// HyperNodeForestBuilder helps in building the forest of hypernodes
+type HyperNodeForestBuilder struct {
+	nodes map[string]*HyperNodeTreeNode // Map to store nodes temporarily
+	roots sets.Set[string]              // Set of root node names
+}
+
+// NewHyperNodeForestBuilder initializes and returns a ForestBuilder
+func NewHyperNodeForestBuilder() *HyperNodeForestBuilder {
+	return &HyperNodeForestBuilder{
+		nodes: make(map[string]*HyperNodeTreeNode),
+		roots: sets.New[string](),
+	}
+}
+
+// insertHyperNodeMember inserts a member into a hypernode and updates the forest
+func insertHyperNodeMember(hntn *HyperNodeTreeNode, hnfb *HyperNodeForestBuilder, memberName string, memberNode *HyperNodeTreeNode, sc *SchedulerCache) {
+	hntn.Members.Insert(memberNode)
+	memberNode.Parent = hntn
+	if hnfb.roots.Has(memberName) {
+		hnfb.roots.Delete(memberName)
+	}
+	sc.HyperNodeListByTier[memberNode.Tier].Insert(memberName)
+
+	sc.HyperNodes[hntn.Name].Union(sc.HyperNodes[memberName])
+}
+
+// addHyperNode adds a HyperNodeTreeNode to the forest based on v1apha1.HyperNode
+// Adding a hypernode can result in updating the entire forest, as it may add new
+// nodes, or link existing nodes - which could demote nodes from roots to child.
+// Because of regex matching, this can trigger ripple effects in the forest.
+func (sc *SchedulerCache) addHyperNode(hn *networktopov1alpha1.HyperNode) {
+	hnfb := sc.hypernodeForestbuilder
+	hntn := NewHyperNodeTreeNode(hn.Name)
+	hntn.HyperNodeRef = hn
+	hnfb.nodes[hn.Name] = hntn
+
+	tierStr := hn.Spec.Tier
+	tier, err := strconv.Atoi(tierStr)
+	if err != nil {
+		klog.Errorf("Failed to convert tier %s to int: %v", tierStr, err)
+		return
+	}
+	hntn.Tier = tier
+	klog.V(4).Infof("Adding hypernode %s to the forest, in %d tier", hn.Name, tier)
+
+	// Iterate over members and add them to the forest
+	for _, member := range hn.Spec.Members {
+		switch member.Type {
+		case networktopov1alpha1.MemberTypeNode:
+			switch member.Selector.Type {
+			case networktopov1alpha1.ExactMatchMemberSelectorType:
+				nodeName := member.Selector.ExactMatch.Name
+				hntn.NodeMatchers.Insert(NewExactMatcher(nodeName))
+				workerNode := sc.Nodes[nodeName]
+				if workerNode == nil {
```

Comment:
Thanks for bring up this scenario. I am aware of it. My hope is that the event handler for Node need to update the HyperNode (I did not implement that logic in the current PR. I was planning to add it later). 

Your proposed solution (using workqueue) should work for Exact match. I wonder how does it work for Regex match. In case of Regex match, when will we know for sure that no future child node will appear -- hence we stop throwing it back to the workqueue?

Reply:
I think a better solution is to update HyperNodeTree when `AddNode` and `DeleteNode`, so that it doesn't matter if Node is not created yet, both `ExactMatch` and `RegexMatch` can be included. Our original intention of designing RegexMatch is actually to serve Node.